Feature/bulma 1

oktaal opened the pull request:
* Server-side pre-rendering
* Upgraded to Bulma 1.0
* Dark mode

oktaal committed:
Upgraded Bulma, added dark mode

oktaal committed:
Merge remote-tracking branch 'origin/develop' into feature/bulma-1

oktaal committed:
Create a fake git user for initial commit (not going to push)

oktaal committed:
Start backend first for SSR

oktaal committed:
fiddly sleep magic

oktaal committed:
Really start backend first?

oktaal committed:
Default value for BACKEND_URL

jgonggrijp commented on {{cookiecutter.slug}}/frontend.angular/src/app/app.component.ts:
```diff
@@ -12,4 +14,21 @@ import { FooterComponent } from './footer/footer.component';
 export class AppComponent {
     title = '{{cookiecutter.project_title}}';
+
+    constructor(@Inject(DOCUMENT) private document: Document, private darkModeService: DarkModeService) {
+        // Using the DOM API to only render on the browser instead of on the server
+        afterRender(() => {
+            const style = this.document.createElement('link');
+            style.rel = 'stylesheet';
+            this.document.head.append(style);
+
+            this.darkModeService.theme$.subscribe(theme => {
+                this.document.documentElement.classList.remove(theme === 'dark' ? 'theme-light' : 'theme-dark');
+                this.document.documentElement.classList.add('theme-' + theme);
+
+                style.href = `${theme}.css`;
+            });
+        });
+    }
```
Please forgive me for the naieve question, but why not just set the theme with a `prefers-color-scheme` media query?

oktaal replied:
Because I cannot change this from the browser (if the user overrides this using the toggle).

jgonggrijp commented on {{cookiecutter.slug}}/frontend.angular/src/app/app.config.server.ts:
```diff
@@ -0,0 +1,12 @@
+const serverConfig: ApplicationConfig = {
+    providers: [
+        provideServerRendering(),
+        { provide: BACKEND_URL, useValue: 'http://localhost:8000' }
```
For my understanding, is this a reference to the Django dev server or is there going to be an additional server running on port 8000? Either way, it seems to me that this will not work equally well in development and in production.

oktaal replied:
It will now start a temporary server when building on production.

jgonggrijp commented on {{cookiecutter.slug}}/frontend.angular/src/app/services/backend.service.ts:
```diff
@@ -32,10 +34,11 @@ export class BackendService {
-            return this.config.get().then(config => config.backendUrl);
-        } else {
-            return Promise.resolve(this.apiUrl);
+            this.apiUrl = this.config.get().then(config => this.backendUrl + config.backendUrl);
```
You are concatenating two properties with the name `backendUrl` here. It is probably correct but it looks suspicious to me. How does it work?

oktaal replied:
Renamed it to `BACKEND_URL_OVERRIDE`.

jgonggrijp commented on {{cookiecutter.slug}}/frontend.angular/src/app/services/dark-mode.service.ts:
```diff
@@ -0,0 +1,87 @@
+        const system =
+            window?.matchMedia && window.matchMedia("(prefers-color-scheme: dark)").matches
+                ? 'dark'
+                : 'light';
```
The return value of `window.matchMedia` can be observed for changes. I suggest doing that, because some people (like myself) configure their operating systems to automatically switch between light mode and dark mode depending on the time of the day.

jgonggrijp commented on {{cookiecutter.slug}}/frontend.angular/src/assets/uu-cdh-dark.svg:
You could preserve the colors by doing an inversion first and then a hue rotation. In fact, you could also achieve that effect with a CSS transform.

oktaal replied:
Not really ideal: `filter: invert() saturate(0)`

<img width="449" alt="image" src="https://github.com/user-attachments/assets/012ca43c-28b6-4d1d-9355-9a6e36c2b149">

But I also don't like the provided logos not being aligned, looking into it.

jgonggrijp replied:
I meant `invert() hue-rotate(180)`. Possibly followed by `brightness(0.875)` and/or `contrast(0.875)`. Try it, it might look better than you expected!

jgonggrijp reviewed (commented):
It seems to work! (The dark mode, at least; I don't know how to check whether server side rendering is working.)

When I first load the page, it automatically selects the mode that I configured my OS (KDE/Fedora) with, which is great. If I then toggle the mode at the OS level, the page does not change along. Making that work would be even greater.

The dark mode CDH logo looks different from the light mode one (beyond brightness, I mean). It looks less good in my opinion, but maybe that is beyond your control.

Otherwise, I like the way it looks. I think you picked good brightness levels and I like the fact that the menu bar stays at the same color across modes. That's a nice playful twist!

oktaal committed:
After feedback @jgonggrijp:
    - Start a temporary server during production build
    - DIfference between BACKEND_URL (http://localhost:8000/api/) and backendPath (/api/)
    - Observe system to allow dynamic switch between dark/light
    - Updated dark logo to align with light logo

oktaal committed:
Clarified BACKEND_URL to BACKEND_URL_OVERRIDE

oktaal commented:
Thanks for the review! I've modified it to address your points.
